contrib/xor-blocks/src/main.rs

```diff
@@ -0,0 +1,193 @@
+// Copyright (c) The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+use std::{
+    env,
+    fmt::{self, Debug, Display, Formatter},
+    fs::{self, File},
+    hash::{BuildHasher, Hasher, RandomState},
+    io::{self, BufReader, BufWriter, Read, Write},
+    path::Path,
+    time::{Duration, Instant},
+};
+
+const MAGIC: [u8; 4] = [0xf9, 0xbe, 0xb4, 0xd9];
+const XOR_FILE_NAME: &str = "xor.dat";
+
+#[derive(Debug)]
+struct Error {
+    msg: String,
+}
+
+impl Error {
+    fn new(msg: String) -> Self {
+        Self { msg }
+    }
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.msg)
+    }
+}
+
+impl From<io::Error> for Error {
+    fn from(value: io::Error) -> Self {
+        Error::new(value.to_string())
```

Comment:
This loses error sources. The proper way to do this is to create a mutable string and then append the sources in a loop with `": "` as a separator. (You could also do cargo-style `"\n\tcaused by: "`).